src/kirby/components/modal/modal-service/modal-service.tns.ts

```diff
@@ -0,0 +1,35 @@
+import { Injectable } from '@angular/core';
+import { ModalDialogService } from 'nativescript-angular/directives/dialogs';
+import { ViewContainerRef } from '@angular/core';
+
+import { ModalComponent } from '~/kirby/components/modal/modal.component';
+import { ModalConfig } from '~/kirby/components/modal/modal-config';
+import { ModalMapService } from './modal-map-service';
+
+@Injectable()
+export class ModalService {
+  constructor(
+    private modalDialogService: ModalDialogService,
+    private modalMapService: ModalMapService
+  ) {}
+
+  public async showModal(config: ModalConfig, vcRef: ViewContainerRef) {
+    const uid = new Date().getTime();
+    config.uid = uid;
+    await this.modalDialogService
+      .showModal(ModalComponent, {
+        viewContainerRef: vcRef,
+        context: config,
+      })
+      .then((callback) => {
+        if (callback) {
+          callback();
+        }
+      });
+    return uid;
+  }
```

Comment:
Excellent point! 

The async/await mechanism simply allows us to do more operations without blocking the interpreter. In this case they are necessary, because of the underlying infrastructure. With this said, async/await is not necessarily related to the return values from the modal.  

However, your comment made me realise I was handling callback logic incorrectly. I would return the data from the modals inside the nested components, which made no sense. I have updated that logic (and the documentation), so it is now possible to create a callback function, in your parent (caller) component, which can (optionally) receive data from a closing modal. Do check the documentation for more insights.

Reply:
Great! :)